pkg/pillar/types/global.go

```diff
@@ -103,312 +106,410 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
-	NetworkTestInterval:       300, // 5 minutes
-	NetworkTestBetterInterval: 0,   // Disabled
-
-	StaleConfigTime:         0, // Don't use stale config
-	DownloadGCTime:          60,
-	VdiskGCTime:             60,
-	DownloadRetryTime:       60,
-	DomainBootRetryTime:     10,
-	Dom0MinDiskUsagePercent: 20,
-	RktGCGracePeriod:        600,
-}
-
-func EnforceGlobalConfigMinimums(newgc GlobalConfig) GlobalConfig {
-
-	if newgc.ConfigInterval < GlobalConfigMinimums.ConfigInterval {
-		log.Warnf("Enforce minimum ConfigInterval received %d; using %d",
-			newgc.ConfigInterval, GlobalConfigMinimums.ConfigInterval)
-		newgc.ConfigInterval = GlobalConfigMinimums.ConfigInterval
-	}
-	if newgc.MetricInterval < GlobalConfigMinimums.MetricInterval {
-		log.Warnf("Enforce minimum MetricInterval received %d; using %d",
-			newgc.MetricInterval, GlobalConfigMinimums.MetricInterval)
-		newgc.MetricInterval = GlobalConfigMinimums.MetricInterval
+func (specMap *ConfigItemSpecMap) AddStringItem(key GlobalSettingKey, defaultString string) {
+	configItem := ConfigItemSpec{
+		ItemType:      ConfigItemTypeString,
+		Key:           string(key),
+		StringDefault: defaultString,
 	}
-	if newgc.ResetIfCloudGoneTime < GlobalConfigMinimums.ResetIfCloudGoneTime {
-		log.Warnf("Enforce minimum XXX received %d; using %d",
-			newgc.ResetIfCloudGoneTime, GlobalConfigMinimums.ResetIfCloudGoneTime)
-		newgc.ResetIfCloudGoneTime = GlobalConfigMinimums.ResetIfCloudGoneTime
+	specMap.GlobalSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddTriStateItem(key GlobalSettingKey, defaultTriState TriState) {
+	configItem := ConfigItemSpec{
+		Key:             string(key),
+		ItemType:        ConfigItemTypeTristate,
+		TriStateDefault: defaultTriState,
 	}
-	if newgc.FallbackIfCloudGoneTime < GlobalConfigMinimums.FallbackIfCloudGoneTime {
-		log.Warnf("Enforce minimum FallbackIfCloudGoneTime received %d; using %d",
-			newgc.FallbackIfCloudGoneTime, GlobalConfigMinimums.FallbackIfCloudGoneTime)
-		newgc.FallbackIfCloudGoneTime = GlobalConfigMinimums.FallbackIfCloudGoneTime
+	specMap.GlobalSettings[key] = configItem
+	log.Debugf("Added tristate item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddAgentSettingStringItem(key AgentSettingKey, defaultString string) {
+	configItem := ConfigItemSpec{
+		ItemType:      ConfigItemTypeString,
+		Key:           string(key),
+		StringDefault: defaultString,
 	}
-	if newgc.MintimeUpdateSuccess < GlobalConfigMinimums.MintimeUpdateSuccess {
-		log.Warnf("Enforce minimum MintimeUpdateSuccess received %d; using %d",
-			newgc.MintimeUpdateSuccess, GlobalConfigMinimums.MintimeUpdateSuccess)
-		newgc.MintimeUpdateSuccess = GlobalConfigMinimums.MintimeUpdateSuccess
+	specMap.AgentSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, item config.ConfigItem) error {
```

Comment:
it would be much better if we take Key, Value as parameters instead of taking "item config.ConfigItem" as a parameter. That is what this function does - It takes String Key, Value, parses them, creates ConfigItemValue and stores it into configMap. Requiring the Key, Value to be passed in as item config.ConfigItem dosn't add any value..